File: leetcode/js/max-increase-to-keep-city-skyline.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
var maxIncreaseKeepingSkyline = function (grid) {
  const n = grid.length;
  let result = 0;

  for (let i = 0; i < n; i++) {
    for (let j = 0; j < n; j++) {
      const cur = grid[i][j];
      let maxRow = cur;
      let maxCol = cur;

      for (let k = 0; k < n; k++) {
        const curRow = grid[k][j];
        const curCol = grid[i][k];

        maxRow = Math.max(maxRow, curRow);
        maxCol = Math.max(maxCol, curCol);
      }

      const min = Math.min(maxRow, maxCol);
      result += Math.abs(cur - min);
    }
  }

  return result;
};

var maxIncreaseKeepingSkyline = function (grid) {
  const n = grid.length;
  const rowMax = new Array(n).fill(0);
  const colMax = new Array(n).fill(0);

  for (let i = 0; i < n; i++) {
    for (let j = 0; j < n; j++) {
      rowMax[i] = Math.max(rowMax[i], grid[i][j]);
      colMax[j] = Math.max(colMax[j], grid[i][j]);
    }
  }

  let result = 0;

  for (let i = 0; i < n; i++) {
    for (let j = 0; j < n; j++) {
      const maxHeight = Math.min(rowMax[i], colMax[j]);
      result += maxHeight - grid[i][j];
    }
  }

  return result;
};

File: leetcode/rust/max-increase-to-keep-city-skyline.rs
impl Solution {
    pub fn max_increase_keeping_skyline(grid: Vec<Vec<i32>>) -> i32 {
        let n = grid.len();
        let mut row_max = vec![0; n];
        let mut col_max = vec![0; n];

        for i in 0..n {
            for j in 0..n {
                row_max[i] = row_max[i].max(grid[i][j]);
                col_max[j] = col_max[j].max(grid[i][j]);
            }
        }

        let mut result = 0;

        for i in 0..n {
            for j in 0..n {
                let max_height = row_max[i].min(col_max[j]);
                result += max_height - grid[i][j];
            }
        }

        result
    }
}
